coriolis/utils.py

```diff
@@ -606,3 +609,198 @@ def create_service(ssh, cmdline, svcname, run_as=None, start=True):
+                            "opt_key": fields[0],
+                        }
+                    )
+            ret.append(
+                {
+                    "type": "option",
+                    "payload": line,
+                    "quoted": quoted,
+                    "option_name": vals[0],
+                    "option_value": opt_vals,
+                }
+            )
+        return ret
+
+    def _validate_value(self, value):
+        if type(value) is not dict:
+            raise ValueError("value was not dict")
+        opt_type = value.get("opt_type")
+        if opt_type not in ("key_val", "single"):
+            raise ValueError("invalid value type %s" % opt_type)
+        if opt_type == "key_val":
+            if "opt_val" not in value or "opt_key" not in value:
+                raise ValueError(
+                        "key_val option type requires "
+                        "opt_key key and opt_val")
+        elif opt_type == "single":
+            if "opt_val" not in value:
+                raise ValueError(
+                        "single option type requires opt_val")
+        else:
+            raise ValueError("unknown option type: %s" % opt_type)
+
+
+    def set_option(self, option, value):
+        """Replaces the value of an option completely
+        """
+        self._validate_value(value)
+        opt_found = False
+        for opt in self._parsed:
+            if opt.get("option_name") == option:
+                opt_found = True
+                opt["option_value"] = value
+                break
+        if not opt_found:
+            self._parsed.append({
+                "type": "option",
+                "quoted": True,
+                "option_name": option,
+                "option_value": [
+                    value
+                ],
+            })
+
+    def append_to_option(self, option, value):
+        """Appends a value to the specified option. If we're passing
+        in a key_val type and the option already exists, the value
+        will be replaced. Options of type "single", if absent from the
+        list, will be appended. If a single value already exists
+        it will be ignored.
+        """
+        self._validate_value(value)
+        opt_found = False
+        for opt in self._parsed:
+            if opt.get("option_name") == option:
+                opt_found = True
+                found = False
+                for val in opt["option_value"]:
+                    if (val["opt_type"] == "key_val" and 
+                            value["opt_type"] == "key_val"):
+                        if str(val["opt_key"]) == str(value["opt_key"]):
+                            val["opt_val"] = value["opt_val"]
+                            found = True
+                    elif (val["opt_type"] == "single" and 
+                            value["opt_type"] == "single"):
+                        if str(val["opt_val"]) == str(value["opt_val"]):
```

Comment:
As far as I can tell if we do find an option of type `single` which is already set to something `!= new_value['opt_val']` then we don't set the new value at all?
This is fine as long as it's intended behavior, but I'd suggest we at the very least `LOG.warn()` about it.

Reply:
Single options are non key/value pair options. For example, a ```key_val``` type is of the form:

```bash
net.ifaces=1
``` 

A single is a simple scalar:
```bash
noapic
noacpi
no_stf_barrier
```

The ```append_to_option()``` function will add an element to an OPTION. An option is one of:

```bash
GRUB_HIDDEN_TIMEOUT
GRUB_CMDLINE_LINUX_DEFAULT
``` 

and so on. So a ```single``` option means adding something like ```noapic``` to one of the grub options.

Now, if a ```single``` option already exists in the grub options you are editing, it will be ignored. If not, it will be **appended**.

So for example, given the option:

```bash
# this already has 2 single flags set
GRUB_CMDLINE_LINUX="noibrs noibpb"
```

calling:

```python
cfg.append_to_option("GRUB_CMDLINE_LINUX", {"opt_type": "single", "opt_val": "noibrs"})
cfg.append_to_option("GRUB_CMDLINE_LINUX", {"opt_type": "single", "opt_val": "noacpi"})
```

will result in:

```bash
GRUB_CMDLINE_LINUX="noibrs noibpb noacpi"
```

because noibrs already existed.

Reply:
Thanks for the explanation, my bad for not getting it the first time.